features/FEATURE_BLE/targets/TARGET_NORDIC/TARGET_NRF5/source/nRF5XPalGattClient.cpp

```diff
@@ -0,0 +1,1616 @@
+    {
+        services_array_t services(
+            evt.params.prim_srvc_disc_rsp.services,
+            evt.params.prim_srvc_disc_rsp.count
+        );
+
+        /**
+         * Implementation of AttReadByGroupTypeResponse that addapt Nordic data
+         * structure.
+         */
+        struct CustomAttReadByGroupTypeResponse : AttReadByGroupTypeResponse {
+            CustomAttReadByGroupTypeResponse(const services_array_t &s, const UUID &u) :
+                AttReadByGroupTypeResponse(), services(s), uuid(u) {
+            }
+
+            virtual size_t size() const
+            {
+                return services.size();
+            }
+
+            virtual attribute_data_t operator[](size_t i) const
+            {
+                attribute_data_t result = {
+                    to_ble_handle_range(services[i].handle_range),
+                    make_ArrayView(uuid.getBaseUUID(), uuid.getLen())
+                };
+                return result;
+            }
+
+            const services_array_t &services;
+            const UUID &uuid;
+        };
+
+        terminate(CustomAttReadByGroupTypeResponse(services, _service_uuid));
+    }
+
+    UUID _service_uuid;
+};
+
+
+/**
+ * Procedure that manage Find Included Services transactions.
+ */
+struct nRF5XGattClient::FindIncludedServicesProcedure : RegularGattProcedure {
+
+    typedef ArrayView<const ble_gattc_service_t> services_array_t;
+
+    FindIncludedServicesProcedure(connection_handle_t connection) :
+        RegularGattProcedure(
+            connection,
+            AttributeOpcode::READ_BY_TYPE_REQUEST,
+            BLE_GATTC_EVT_REL_DISC_RSP
+        ) { }
+
+    ble_error_t start(attribute_handle_range_t service_range)
+    {
+        ble_gattc_handle_range_t range = to_nordic_handle_range(service_range);
+        uint32_t err = sd_ble_gattc_relationships_discover(
+            connection_handle, &range
+        );
+
+        return convert_sd_error(err);
+    }
+
+    virtual void do_handle(const ble_gattc_evt_t &evt)
+    {
+        // recompose the message into a raw att read by type response.
+        const ble_gattc_evt_rel_disc_rsp_t &resp = evt.params.rel_disc_rsp;
+
+        bool contain_short_uuid =
+            (resp.includes[0].included_srvc.uuid.type == BLE_UUID_TYPE_BLE);
+
+        uint8_t element_size = 6;
+        if (contain_short_uuid) {
+            element_size += 2;
+        }
+
+        // It would be more efficient to use an API like alloca but it is
+        // unavailable and unsafe. Another alternative would be to have a fixed
+        // size stack buffer since the size of the MTU is fixed however once new
+        // softdevices lands could would have to be rewritten because they support
+        // variable MTU size.
+        size_t buffer_size = element_size * resp.count;
+        uint8_t* buffer = new(std::nothrow) uint8_t[buffer_size];
+        if (!buffer) {
+            abort();
+        }
+
+        uint8_t *it = buffer;
```

Comment:
is this supposed to continue past abort()?

Reply:
Nope, good catch!

Reply:
There are a few more aborts that continue, you might want to have a quick look